index.bs

```diff
@@ -295,50 +295,61 @@ partial interface NeuralNetworkContext {
                  optional OperandLayout layout = "nchw");
 };
 </script>
-
 <div algorithm=conv2d>
     **Arguments:**
-        - *input*: an {{Operand}}. The input 4-D tensor. The logical shape
-            is interpreted according to the value of *layout*.
-        - *filter*: an {{Operand}}. The filter 4-D tensor. The logical shape
-            is interpreted according to the value of *layout*.
-        - *padding*: a sequence of {{long}} of length 4. The padding for the
-            beginning and ending along each spatial dimension of *input*,
-            [beginning_height, ending_height, beginning_width, ending_width].
-        - *strides*: a sequence of {{long}} of length 2. The stride of the
-            sliding window for each spatial dimension of *input*,
-            [stride_height, stirde_width].
-        - *dilations*: a sequence of {{long}} of length 2. The dilation factor
-            for each spatial dimension of *input*, [dilation_height,
-            dilation_width].
+        - *input*: an {{Operand}}. The input 4-D tensor. The logical shape is
+            interpreted according to the value of *layout*.
+        - *filter*: an {{Operand}}. The filter 4-D or 5-D tensor. The logical
+            shape is interpreted according to the value of *layout* and
+            *groups*.
+        - *padding*: an optional sequence of {{long}} of length 4. The padding
+            for the beginning and ending along each spatial dimension of
+            *input*, [beginning_height, ending_height, beginning_width,
+            ending_width]. If not present, the values are assumed to be
+            [0,0,0,0].
+        - *strides*: an optional sequence of {{long}} of length 2. The stride of
+            the sliding window for each spatial dimension of *input*,
+            [stride_height, stride_width]. If not present, the values are
+            assumed to be [1,1].
+        - *dilations*: an optional sequence of {{long}} of length 2. The
+            dilation factor for each spatial dimension of *input*,
+            [dilation_height, dilation_width]. If not present, the values are
+            assumed to be [1,1].
+        - *groups*" an optional {{long}}. The number of groups that input
+            channels and output channels are divided into, default to 1.
         - *layout*: an optional {{OperandLayout}} with value as "nchw" or
             "nhwc". The default value is "nchw". This argument specifies the
             layout format of the input, output and filter tensor.
             Specifically,
 
             "nchw":
                 - input tensor: [batches, input_channels, height, width]
-                - filter tensor: [output_channels, input_channels, height,
-                    width]
+                - filter tensor: [output_channels, input_channels,
+                    height, width] when *groups* is equal to 1;
+                    [groups, output_channels/groups, input_channels/groups,
```

Comment:
For grouped convolution, the filter shape should still be 4D of `[groups, input_channels/groups, H, W]` because the number of output channels will be derived by the process to be the number of groups, so the output tensor shape is `[batches, groups, H, W]`.

For `nhwc` (*channel last* format), the filter shape is `[H, W, groups, input_channels/groups]`.

Note that to implement depthwise in MobileNet, the group count is set to the number of input channels, so if the input channels is, say, 3, the filter shape will be `[3, 1, H, W]` and the output `[batches, 3, H, W]` accordingly (for 'nchw' layout).

Reply:
@huningxin for `nchw` I believe we are saying the same thing. When group count is > 1, we are basically breaking up the filter that used to be of `input_channels` depth into groups. But because the total number of channels across all the groups must still be equal to `input_channels` , the filter shape therefore is `[groups, input_channels/groups, H, W]` as the total number of elements in the tensor will still be `groups * input_channels/groups * H * W` = `input_channels * H * W`. The `output_channels` is irrelevant in this case because it's *depthwise*, the number of output channels is only relevant when there is just 1 group of filter channels in use. Hope it makes sense. 

As for `nhwc`, it is `[H, W, groups, input_channels/groups]` because again the total number of tensor elements of the filter across all groups is still `H * W * input_channels'.

Reply:
> The `output_channels` is irrelevant in this case because it's _depthwise_, the number of output channels is only relevant when there is just 1 group of filter channels in use. Hope it makes sense.

Thanks for the explanation @wchao1115 . I have no questions about _depthwise_ case. But for general cases, the filter shape as `[groups, input_channels/groups, H, W]` seems not work. For instance, in [AlexNet ONNX model](https://github.com/onnx/models/tree/master/vision/classification/alexnet), the conv with output "conv4_1" has `input_channels == 384` and `groups == 2`. According to `[groups, input_channels/groups, H, W]` definition, the filter shape would be `[2, 192, 3, 3]`. However,  the actual filter shape is `[384, 192, 3, 3]` where the `output channels == 384`. So I propose to define filter shape as `[output_channels, input_channels/groups, H, W]`.

I post a netron screenshot for your reference. Hope it helps.
![image](https://user-images.githubusercontent.com/1005673/83850476-91dc4200-a743-11ea-8598-f7cc569bf3b2.png)

Reply:
I checked with our implementation in DirectML and you're right @huningxin. Depthwise is just a special case when `groups` = `input_channels` = `output_channels`. The more generalized use case of grouped convolution has the filter shape for `nchw`  and `nhwc` as `[output_channels, input_channels/groups, H, W]` and `[H, W, input_channels/groups, output_channels]` respectively. Does that sound right? Thanks for the sample on AlexNet.

Maybe we should also call out the depthwise special case in a note section to be more specific about how it can be achieved using grouped convolution.

Reply:
@wchao1115 , thanks for the checking. I pushed a new commit that changes the filter tensor to 4D and adds depthwise case in a note. Please take another look.

Reply:
Looks good @huningxin. I just have a nit suggestion in the commit to make the note section a bit more clear.